pkg/sql/gcjob_test/gcjob_test_test: TestSchemaChangeGCJob failed

pkg/sql/gcjob_test/gcjob_test_test.TestSchemaChangeGCJob [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11267923?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11267923?buildTab=artifacts#/) on master @ [7fe93d78f572799da9a6c428768845962ceab610](https://github.com/cockroachdb/cockroach/commits/7fe93d78f572799da9a6c428768845962ceab610):


```
I230810 15:58:57.774065 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221529  didnt refresh descriptor: 36 lease: node unavailable; try another peer
I230810 15:58:57.774232 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221530  didnt refresh descriptor: 12 lease: node unavailable; try another peer
I230810 15:58:57.774465 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221531  didnt refresh descriptor: 28 lease: node unavailable; try another peer
I230810 15:58:57.774635 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221532  didnt refresh descriptor: 7 lease: node unavailable; try another peer
I230810 15:58:57.774849 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221533  didnt refresh descriptor: 40 lease: node unavailable; try another peer
I230810 15:58:57.775123 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221534  didnt refresh descriptor: 39 lease: node unavailable; try another peer
I230810 15:58:57.775338 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221535  didnt refresh descriptor: 47 lease: node unavailable; try another peer
I230810 15:58:57.775527 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221536  didnt refresh descriptor: 44 lease: node unavailable; try another peer
I230810 15:58:57.775746 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221537  didnt refresh descriptor: 20 lease: node unavailable; try another peer
I230810 15:58:57.775916 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221538  didnt refresh descriptor: 104 lease: node unavailable; try another peer
I230810 15:58:57.776159 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221539  didnt refresh descriptor: 59 lease: node unavailable; try another peer
I230810 15:58:57.776336 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221540  didnt refresh descriptor: 62 lease: node unavailable; try another peer
I230810 15:58:57.776566 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221541  didnt refresh descriptor: 51 lease: node unavailable; try another peer
I230810 15:58:57.776807 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221542  didnt refresh descriptor: 106 lease: node unavailable; try another peer
I230810 15:58:57.776924 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221543  didnt refresh descriptor: 52 lease: node unavailable; try another peer
I230810 15:58:57.777123 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221544  didnt refresh descriptor: 50 lease: node unavailable; try another peer
I230810 15:58:57.777276 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221545  didnt refresh descriptor: 21 lease: node unavailable; try another peer
I230810 15:58:57.777406 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221546  didnt refresh descriptor: 11 lease: node unavailable; try another peer
I230810 15:58:57.777566 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221547  didnt refresh descriptor: 30 lease: node unavailable; try another peer
I230810 15:58:57.777737 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221548  didnt refresh descriptor: 34 lease: node unavailable; try another peer
I230810 15:58:57.777942 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221549  didnt refresh descriptor: 60 lease: node unavailable; try another peer
I230810 15:58:57.778120 2356 sql/catalog/lease/lease.go:1336  [T1,n1] 221550  didnt refresh descriptor: 23 lease: node unavailable; try another peer
W230810 15:58:57.778466 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221551  failed to flush statement statistics: flush statement 16341497805840073494's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:57.797550 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221552  failed to flush statement statistics: flush statement 1233060761775477877's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:57.816965 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221553  failed to flush statement statistics: flush statement 5219567332652146454's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:57.835713 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221554  failed to flush statement statistics: flush statement 17675461270115263172's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:57.854503 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221555  failed to flush statement statistics: flush statement 3942184553683936799's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:57.873536 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221556  failed to flush statement statistics: flush statement 12429994604490846754's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:57.892029 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221557  failed to flush statement statistics: flush statement 18014002614525947250's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:57.910942 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221558  failed to flush statement statistics: flush statement 6782457755092058661's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:57.930082 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221559  failed to flush statement statistics: flush statement 3152240207279189612's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:57.948504 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221560  failed to flush statement statistics: flush statement 14884669397301568260's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:57.967186 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221561  failed to flush statement statistics: flush statement 16885529018338426180's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:57.986505 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221562  failed to flush statement statistics: flush statement 8157900015381273945's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:58.005664 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221563  failed to flush statement statistics: flush statement 357176498942065901's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:58.023948 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221564  failed to flush statement statistics: flush statement 14587664299696923629's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:58.042567 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221565  failed to flush statement statistics: flush statement 21943780307261950's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:58.061349 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221566  failed to flush statement statistics: flush statement 8565716383475309113's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:58.080316 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221567  failed to flush statement statistics: flush statement 6348215287580425483's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:58.099489 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221568  failed to flush statement statistics: flush statement 6348203192952515288's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:58.117971 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221569  failed to flush statement statistics: flush statement 10636516999895996107's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:58.136892 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221570  failed to flush statement statistics: flush statement 1087927460085071165's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:58.154547 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221571  failed to flush statement statistics: flush statement 13953538446758315263's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:58.173468 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221572  failed to flush statement statistics: flush statement 2756509097058173994's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:58.192093 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221573  failed to flush statement statistics: flush statement 3419408685265610513's statistics: insert-stmt-stats: node unavailable; try another peer
W230810 15:58:58.210747 10530406 sql/sqlstats/persistedsqlstats/flush.go:164  [T1,n1] 221574  failed to flush statement statistics: flush statement 18387747675008746556's statistics: insert-stmt-stats: node unavailable; try another peer
I230810 15:58:58.230992 14 server/server_controller_orchestration.go:268  [T1,n1] 221575  server controller shutting down ungracefully
I230810 15:58:58.231167 14 server/server_controller_orchestration.go:279  [T1,n1] 221576  waiting for tenant servers to report stopped
W230810 15:58:58.234991 14 server/server_sql.go:1689  [T1,n1] 221577  server shutdown without a prior graceful drain
--- FAIL: TestSchemaChangeGCJob (3431.09s)
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSchemaChangeGCJob.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-30506

--- a/pkg/sql/gcjob_test/gc_job_test.go
+++ b/pkg/sql/gcjob_test/gc_job_test.go
@@ -53,218 +53,228 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+type DropItem int
+
+const (
+	INDEX = iota
+	TABLE
+	DATABASE
+)
+
+type TTLTime int
+
+const (
+	PAST   = iota // An item was supposed to be GC already.
+	SOON          // An item will be GC'd soon.
+	FUTURE        // An item should not be GC'd during this test.
+)
+
 // TODO(pbardea): Add more testing around the timer calculations.
 func TestSchemaChangeGCJob(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-
-	type DropItem int
-	const (
-		INDEX = iota
-		TABLE
-		DATABASE
-	)
-
-	type TTLTime int
-	const (
-		PAST   = iota // An item was supposed to be GC already.
-		SOON          // An item will be GC'd soon.
-		FUTURE        // An item should not be GC'd during this test.
-	)
+	defer log.Scope(t).Close(t)
 
 	for _, dropItem := range []DropItem{INDEX, TABLE, DATABASE} {
 		for _, ttlTime := range []TTLTime{PAST, SOON, FUTURE} {
-			blockGC := make(chan struct{}, 1)
-			params := base.TestServerArgs{}
-			params.ScanMaxIdleTime = time.Millisecond
-			params.Knobs.JobsTestingKnobs = jobs.NewTestingKnobsWithShortIntervals()
-			params.Knobs.GCJob = &sql.GCJobTestingKnobs{
-				RunBeforePerformGC: func(_ jobspb.JobID) error {
-					<-blockGC
-					return nil
-				},
-			}
-			s, db, kvDB := serverutils.StartServer(t, params)
-			ctx := context.Background()
-			defer s.Stopper().Stop(ctx)
-			sqlDB := sqlutils.MakeSQLRunner(db)
-
-			sqlDB.Exec(t, `SET CLUSTER SETTING sql.gc_job.wait_for_gc.interval = '1s';`)
-			// Refresh protected timestamp cache immediately to make MVCC GC queue to
-			// process GC immediately.
-			sqlDB.Exec(t, `SET CLUSTER SETTING kv.protectedts.poll_interval = '1s';`)
-
-			jobRegistry := s.JobRegistry().(*jobs.Registry)
-
-			sqlDB.Exec(t, "CREATE DATABASE my_db")
-			sqlDB.Exec(t, "USE my_db")
-			sqlDB.Exec(t, "CREATE TABLE my_table (a int primary key, b int, index (b))")
-			sqlDB.Exec(t, "CREATE TABLE my_other_table (a int primary key, b int, index (b))")
-			if ttlTime == SOON {
-				sqlDB.Exec(t, "ALTER TABLE my_table CONFIGURE ZONE USING gc.ttlseconds = 1")
-				sqlDB.Exec(t, "ALTER TABLE my_other_table CONFIGURE ZONE USING gc.ttlseconds = 1")
-			}
-
-			myDBID := descpb.ID(bootstrap.TestingUserDescID(4))
-			myTableID := descpb.ID(bootstrap.TestingUserDescID(6))
-			myOtherTableID := descpb.ID(bootstrap.TestingUserDescID(7))
-
-			var myTableDesc *tabledesc.Mutable
-			var myOtherTableDesc *tabledesc.Mutable
-			if err := sql.TestingDescsTxn(ctx, s, func(ctx context.Context, txn isql.Txn, col *descs.Collection) error {
-				myImm, err := col.ByID(txn.KV()).Get().Table(ctx, myTableID)
-				if err != nil {
-					return err
-				}
-				myTableDesc = tabledesc.NewBuilder(myImm.TableDesc()).BuildExistingMutableTable()
-				myOtherImm, err := col.ByID(txn.KV()).Get().Table(ctx, myOtherTableID)
-				if err != nil {
-					return err
-				}
-				myOtherTableDesc = tabledesc.NewBuilder(myOtherImm.TableDesc()).BuildExistingMutableTable()
-				return nil
-			}); err != nil {
-				t.Fatal(err)
-			}
-
-			// Start the job that drops an index.
-			dropTime := timeutil.Now().UnixNano()
-			if ttlTime == PAST {
-				dropTime = 1
-			}
-			var details jobspb.SchemaChangeGCDetails
-			var expectedRunningStatus string
-			switch dropItem {
-			case INDEX:
-				details = jobspb.SchemaChangeGCDetails{
-					Indexes: []jobspb.SchemaChangeGCDetails_DroppedIndex{
-						{
-							IndexID:  descpb.IndexID(2),
-							DropTime: dropTime,
-						},
-					},
-					ParentID: myTableID,
-				}
-				myTableDesc.SetPublicNonPrimaryIndexes([]descpb.IndexDescriptor{})
-				expectedRunningStatus = "deleting data"
-			case TABLE:
-				details = jobspb.SchemaChangeGCDetails{
-					Tables: []jobspb.SchemaChangeGCDetails_DroppedID{
-						{
-							ID:       myTableID,
-							DropTime: dropTime,
-						},
-					},
-				}
-				myTableDesc.State = descpb.DescriptorState_DROP
-				myTableDesc.DropTime = dropTime
-				expectedRunningStatus = "deleting data"
-			case DATABASE:
-				details = jobspb.SchemaChangeGCDetails{
-					Tables: []jobspb.SchemaChangeGCDetails_DroppedID{
-						{
-							ID:       myTableID,
-							DropTime: dropTime,
-						},
-						{
-							ID:       myOtherTableID,
-							DropTime: dropTime,
-						},
-					},
-					ParentID: myDBID,
-				}
-				myTableDesc.State = descpb.DescriptorState_DROP
-				myTableDesc.DropTime = dropTime
-				myOtherTableDesc.State = descpb.DescriptorState_DROP
-				myOtherTableDesc.DropTime = dropTime
-				expectedRunningStatus = "deleting data"
-			}
-
-			if err := kvDB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
-				b := txn.NewBatch()
-				descKey := catalogkeys.MakeDescMetadataKey(keys.SystemSQLCodec, myTableID)
-				descDesc := myTableDesc.DescriptorProto()
-				b.Put(descKey, descDesc)
-				descKey2 := catalogkeys.MakeDescMetadataKey(keys.SystemSQLCodec, myOtherTableID)
-				descDesc2 := myOtherTableDesc.DescriptorProto()
-				b.Put(descKey2, descDesc2)
-				return txn.Run(ctx, b)
-			}); err != nil {
-				t.Fatal(err)
-			}
-
-			jobRecord := jobs.Record{
-				Description:   "GC test",
-				Username:      username.TestUserName(),
-				DescriptorIDs: descpb.IDs{myTableID},
-				Details:       details,
-				Progress:      jobspb.SchemaChangeGCProgress{},
-				RunningStatus: sql.RunningStatusWaitingGC,
-				NonCancelable: true,
-			}
-
-			// The job record that will be used to lookup this job.
-			lookupJR := jobs.Record{
-				Description:   "GC test",
-				Username:      username.TestUserName(),
-				DescriptorIDs: descpb.IDs{myTableID},
-				Details:       details,
-			}
-
-			job, err := jobs.TestingCreateAndStartJob(ctx, jobRegistry, s.InternalDB().(isql.DB), jobRecord)
-			if err != nil {
-				t.Fatal(err)
-			}
-
-			// Check that the job started.
-			jobIDStr := strconv.Itoa(int(job.ID()))
-			if err := jobutils.VerifyRunningSystemJob(t, sqlDB, 0, jobspb.TypeSchemaChangeGC, sql.RunningStatusWaitingGC, lookupJR); err != nil {
-				t.Fatal(err)
-			}
-
-			if ttlTime != FUTURE {
-				// Check that the job eventually blocks right before performing GC, due to the testing knob.
-				sqlDB.CheckQueryResultsRetry(
-					t,
-					fmt.Sprintf("SELECT status, running_status FROM [SHOW JOBS] WHERE job_id = %s", jobIDStr),
-					[][]string{{"running", expectedRunningStatus}})
-			}
-			blockGC <- struct{}{}
-
-			if ttlTime == FUTURE {
-				time.Sleep(500 * time.Millisecond)
-			} else {
-				sqlDB.CheckQueryResultsRetry(t, fmt.Sprintf("SELECT status FROM [SHOW JOBS] WHERE job_id = %s", jobIDStr), [][]string{{"succeeded"}})
-				if err := jobutils.VerifySystemJob(t, sqlDB, 0, jobspb.TypeSchemaChangeGC, jobs.StatusSucceeded, lookupJR); err != nil {
-					t.Fatal(err)
-				}
-			}
-
-			if err := sql.TestingDescsTxn(ctx, s, func(ctx context.Context, txn isql.Txn, col *descs.Collection) error {
-				myImm, err := col.ByID(txn.KV()).Get().Table(ctx, myTableID)
-				if err != nil {
-					if ttlTime != FUTURE && (dropItem == TABLE || dropItem == DATABASE) {
-						// We dropped the table, so expect it to not be found.
-						require.EqualError(t, err, fmt.Sprintf(`relation "[%d]" does not exist`, myTableID))
-						return nil
-					}
-					return err
-				}
-				myTableDesc = tabledesc.NewBuilder(myImm.TableDesc()).BuildExistingMutableTable()
-				myOtherImm, err := col.ByID(txn.KV()).Get().Table(ctx, myOtherTableID)
-				if err != nil {
-					if ttlTime != FUTURE && dropItem == DATABASE {
-						// We dropped the entire database, so expect none of the tables to be found.
-						require.EqualError(t, err, fmt.Sprintf(`relation "[%d]" does not exist`, myOtherTableID))
-						return nil
-					}
-					return err
-				}
-				myOtherTableDesc = tabledesc.NewBuilder(myOtherImm.TableDesc()).BuildExistingMutableTable()
-				return nil
-			}); err != nil {
-				t.Fatal(err)
-			}
+			// NB: The inner body of this loop has been extracted into a function to
+			// ensure defer statements (namely testserver.Stop) is executed at the
+			// end of each loop rather than the end of each test.
+			doTestSchemaChangeGCJob(t, dropItem, ttlTime)
 		}
+	}
+}
+
+func doTestSchemaChangeGCJob(t *testing.T, dropItem DropItem, ttlTime TTLTime) {
+	blockGC := make(chan struct{}, 1)
+	params := base.TestServerArgs{}
+	params.ScanMaxIdleTime = time.Millisecond
+	params.Knobs.JobsTestingKnobs = jobs.NewTestingKnobsWithShortIntervals()
+	params.Knobs.GCJob = &sql.GCJobTestingKnobs{
+		RunBeforePerformGC: func(_ jobspb.JobID) error {
+			<-blockGC
+			return nil
+		},
+	}
+	s, db, kvDB := serverutils.StartServer(t, params)
+	ctx := context.Background()
+	defer s.Stopper().Stop(ctx)
+	sqlDB := sqlutils.MakeSQLRunner(db)
+
+	sqlDB.Exec(t, `SET CLUSTER SETTING sql.gc_job.wait_for_gc.interval = '1s';`)
+	// Refresh protected timestamp cache immediately to make MVCC GC queue to
+	// process GC immediately.
+	sqlDB.Exec(t, `SET CLUSTER SETTING kv.protectedts.poll_interval = '1s';`)
+
+	jobRegistry := s.JobRegistry().(*jobs.Registry)
+
+	sqlDB.Exec(t, "CREATE DATABASE my_db")
+	sqlDB.Exec(t, "USE my_db")
+	sqlDB.Exec(t, "CREATE TABLE my_table (a int primary key, b int, index (b))")
+	sqlDB.Exec(t, "CREATE TABLE my_other_table (a int primary key, b int, index (b))")
+	if ttlTime == SOON {
+		sqlDB.Exec(t, "ALTER TABLE my_table CONFIGURE ZONE USING gc.ttlseconds = 1")
+		sqlDB.Exec(t, "ALTER TABLE my_other_table CONFIGURE ZONE USING gc.ttlseconds = 1")
+	}
+
+	myDBID := descpb.ID(bootstrap.TestingUserDescID(4))
+	myTableID := descpb.ID(bootstrap.TestingUserDescID(6))
+	myOtherTableID := descpb.ID(bootstrap.TestingUserDescID(7))
+
+	var myTableDesc *tabledesc.Mutable
+	var myOtherTableDesc *tabledesc.Mutable
+	if err := sql.TestingDescsTxn(ctx, s, func(ctx context.Context, txn isql.Txn, col *descs.Collection) error {
+		myImm, err := col.ByID(txn.KV()).Get().Table(ctx, myTableID)
+		if err != nil {
+			return err
+		}
+		myTableDesc = tabledesc.NewBuilder(myImm.TableDesc()).BuildExistingMutableTable()
+		myOtherImm, err := col.ByID(txn.KV()).Get().Table(ctx, myOtherTableID)
+		if err != nil {
+			return err
+		}
+		myOtherTableDesc = tabledesc.NewBuilder(myOtherImm.TableDesc()).BuildExistingMutableTable()
+		return nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	// Start the job that drops an index.
+	dropTime := timeutil.Now().UnixNano()
+	if ttlTime == PAST {
+		dropTime = 1
+	}
+	var details jobspb.SchemaChangeGCDetails
+	var expectedRunningStatus string
+	switch dropItem {
+	case INDEX:
+		details = jobspb.SchemaChangeGCDetails{
+			Indexes: []jobspb.SchemaChangeGCDetails_DroppedIndex{
+				{
+					IndexID:  descpb.IndexID(2),
+					DropTime: dropTime,
+				},
+			},
+			ParentID: myTableID,
+		}
+		myTableDesc.SetPublicNonPrimaryIndexes([]descpb.IndexDescriptor{})
+		expectedRunningStatus = "deleting data"
+	case TABLE:
+		details = jobspb.SchemaChangeGCDetails{
+			Tables: []jobspb.SchemaChangeGCDetails_DroppedID{
+				{
+					ID:       myTableID,
+					DropTime: dropTime,
+				},
+			},
+		}
+		myTableDesc.State = descpb.DescriptorState_DROP
+		myTableDesc.DropTime = dropTime
+		expectedRunningStatus = "deleting data"
+	case DATABASE:
+		details = jobspb.SchemaChangeGCDetails{
+			Tables: []jobspb.SchemaChangeGCDetails_DroppedID{
+				{
+					ID:       myTableID,
+					DropTime: dropTime,
+				},
+				{
+					ID:       myOtherTableID,
+					DropTime: dropTime,
+				},
+			},
+			ParentID: myDBID,
+		}
+		myTableDesc.State = descpb.DescriptorState_DROP
+		myTableDesc.DropTime = dropTime
+		myOtherTableDesc.State = descpb.DescriptorState_DROP
+		myOtherTableDesc.DropTime = dropTime
+		expectedRunningStatus = "deleting data"
+	}
+
+	if err := kvDB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+		b := txn.NewBatch()
+		descKey := catalogkeys.MakeDescMetadataKey(keys.SystemSQLCodec, myTableID)
+		descDesc := myTableDesc.DescriptorProto()
+		b.Put(descKey, descDesc)
+		descKey2 := catalogkeys.MakeDescMetadataKey(keys.SystemSQLCodec, myOtherTableID)
+		descDesc2 := myOtherTableDesc.DescriptorProto()
+		b.Put(descKey2, descDesc2)
+		return txn.Run(ctx, b)
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	jobRecord := jobs.Record{
+		Description:   "GC test",
+		Username:      username.TestUserName(),
+		DescriptorIDs: descpb.IDs{myTableID},
+		Details:       details,
+		Progress:      jobspb.SchemaChangeGCProgress{},
+		RunningStatus: sql.RunningStatusWaitingGC,
+		NonCancelable: true,
+	}
+
+	// The job record that will be used to lookup this job.
+	lookupJR := jobs.Record{
+		Description:   "GC test",
+		Username:      username.TestUserName(),
+		DescriptorIDs: descpb.IDs{myTableID},
+		Details:       details,
+	}
+
+	job, err := jobs.TestingCreateAndStartJob(ctx, jobRegistry, s.InternalDB().(isql.DB), jobRecord)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Check that the job started.
+	jobIDStr := strconv.Itoa(int(job.ID()))
+	if err := jobutils.VerifyRunningSystemJob(t, sqlDB, 0, jobspb.TypeSchemaChangeGC, sql.RunningStatusWaitingGC, lookupJR); err != nil {
+		t.Fatal(err)
+	}
+
+	if ttlTime != FUTURE {
+		// Check that the job eventually blocks right before performing GC, due to the testing knob.
+		sqlDB.CheckQueryResultsRetry(
+			t,
+			fmt.Sprintf("SELECT status, running_status FROM [SHOW JOBS] WHERE job_id = %s", jobIDStr),
+			[][]string{{"running", expectedRunningStatus}})
+	}
+	blockGC <- struct{}{}
+
+	if ttlTime == FUTURE {
+		time.Sleep(500 * time.Millisecond)
+	} else {
+		sqlDB.CheckQueryResultsRetry(t, fmt.Sprintf("SELECT status FROM [SHOW JOBS] WHERE job_id = %s", jobIDStr), [][]string{{"succeeded"}})
+		if err := jobutils.VerifySystemJob(t, sqlDB, 0, jobspb.TypeSchemaChangeGC, jobs.StatusSucceeded, lookupJR); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	if err := sql.TestingDescsTxn(ctx, s, func(ctx context.Context, txn isql.Txn, col *descs.Collection) error {
+		myImm, err := col.ByID(txn.KV()).Get().Table(ctx, myTableID)
+		if err != nil {
+			if ttlTime != FUTURE && (dropItem == TABLE || dropItem == DATABASE) {
+				// We dropped the table, so expect it to not be found.
+				require.EqualError(t, err, fmt.Sprintf(`relation "[%d]" does not exist`, myTableID))
+				return nil
+			}
+			return err
+		}
+		myTableDesc = tabledesc.NewBuilder(myImm.TableDesc()).BuildExistingMutableTable()
+		myOtherImm, err := col.ByID(txn.KV()).Get().Table(ctx, myOtherTableID)
+		if err != nil {
+			if ttlTime != FUTURE && dropItem == DATABASE {
+				// We dropped the entire database, so expect none of the tables to be found.
+				require.EqualError(t, err, fmt.Sprintf(`relation "[%d]" does not exist`, myOtherTableID))
+				return nil
+			}
+			return err
+		}
+		myOtherTableDesc = tabledesc.NewBuilder(myOtherImm.TableDesc()).BuildExistingMutableTable()
+		return nil
+	}); err != nil {
+		t.Fatal(err)
 	}
 }
 

